magma/wire_container.py

```diff
@@ -169,7 +170,12 @@ def bit(self):
 class Wireable:
     def __init__(self):
         self._wire = Wire(self)
-        self._when_context = get_curr_when_block()
+        if isinstance(self.name, ArrayRef):
```

Comment:
I think the derived class (Array) should handle this rather than the parent class (it's kind of backwards). I would propose exposing a (private) method `_set_when_context()` and when Array creates it's children, it should `_set_when_context(self._when_context)`. The main reason I prefer it this way is that the laziness of the array should be fully encapsulated (as much as possible) to the internals of Array, so no other functions or classes should really have to consider that implementation detail.

There is still the problem (either the way you have it here or the way I propose^ that further nested types don't get the appropriate context.

Reply:
For dealing with nesting, we can add a set_when_context API. For arrays, it sets the wireable when context which propagates it to children when they're created.  For tuples, it eagerly calls it on the children (will change in Tuple2, which will share the same code with Array).  For other types, they are leafs that will inherit the standard set_when_context API from Wireable.